### PR TITLE
fix(components): [tabs] optimize SSR

### LINKED ISSUE #15088: [Component] [tabs, tab-pane] El-tabs not support SSR.
Bug Type: **`Component`**

## Environment

- Vue Version: `3.4.0-alpha.4`
- Element Plus Version: `2.4.3`
- Browser / OS: ` Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-tabs`
- `el-tab-pane`

### Reproduction Link

[Github Repo](https://github.com/element-plus/element-plus-nuxt/issues/78)

### Steps to reproduce

https://github.com/element-plus/element-plus-nuxt/issues/78

Directly import `el-tabs` in this example.

### What is Expected?

It can render in server side.

### What is actually happening?

Only render in client.

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by makedopamine**
fixes #15088

[before](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLWJ1dHRvbiBAY2xpY2s9XCJhcHAubW91bnQoJyNjb250YWluZXInKVwiPmh5ZHJhdGU8L2VsLWJ1dHRvbj5cbiAgPGRpdiBpZD1cImNvbnRhaW5lclwiPjwvZGl2PlxuPC90ZW1wbGF0ZT5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyBjcmVhdGVTU1JBcHAsIGgsIG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IHJlbmRlclRvU3RyaW5nIH0gZnJvbSAndnVlL3NlcnZlci1yZW5kZXJlcidcbmltcG9ydCB7IEVsVGFicywgRWxUYWJQYW5lIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCBhcHAgPSBjcmVhdGVTU1JBcHAoe1xuICByZW5kZXIoKSB7XG4gICAgY29uc3QgdGFiUGFuZXMgPSBbJ2ZvbycsICdiYXInXS5tYXAoKHYpID0+XG4gICAgICBoKFxuICAgICAgICBFbFRhYlBhbmUsXG4gICAgICAgIHtcbiAgICAgICAgICBsYWJlbDogdixcbiAgICAgICAgICBuYW1lOiB2LFxuICAgICAgICB9LFxuICAgICAgICB7XG4gICAgICAgICAgZGVmYXVsdDogKCkgPT4gdixcbiAgICAgICAgfVxuICAgICAgKVxuICAgIClcblxuICAgIHJldHVybiBoKFxuICAgICAgRWxUYWJzLFxuICAgICAge1xuICAgICAgICBtb2RlbFZhbHVlOiAnYmFyJyxcbiAgICAgIH0sXG4gICAgICB7IGRlZmF1bHQ6ICgpID0+IHRhYlBhbmVzIH1cbiAgICApXG4gIH0sXG59KVxuXG5vbk1vdW50ZWQoKCkgPT4ge1xuICByZW5kZXJUb1N0cmluZyhhcHApLnRoZW4oKGh0bWwpID0+IHtcbiAgICBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCcjY29udGFpbmVyJykuaW5uZXJIVE1MID0gaHRtbFxuICB9KVxufSlcbjwvc2NyaXB0PlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS92dWUvZGlzdC92dWUucnVudGltZS5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwidnVlL3NlcnZlci1yZW5kZXJlclwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zZXJ2ZXItcmVuZGVyZXJAbGF0ZXN0L2Rpc3Qvc2VydmVyLXJlbmRlcmVyLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJAdnVlL3NoYXJlZFwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJAdnVlL3J1bnRpbWUtZG9tXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtZG9tQGxhdGVzdC9kaXN0L3J1bnRpbWUtZG9tLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJAdnVlL3J1bnRpbWUtY29yZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWNvcmVAbGF0ZXN0L2Rpc3QvcnVudGltZS1jb3JlLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJAdnVlL3JlYWN0aXZpdHlcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcmVhY3Rpdml0eUBsYXRlc3QvZGlzdC9yZWFjdGl2aXR5LmVzbS1idW5kbGVyLmpzXCJ9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIl9vIjp7fX0=)

[after](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLWJ1dHRvbiBAY2xpY2s9XCJhcHAubW91bnQoJyNjb250YWluZXInKVwiPmh5ZHJhdGU8L2VsLWJ1dHRvbj5cbiAgPGRpdiBpZD1cImNvbnRhaW5lclwiPjwvZGl2PlxuPC90ZW1wbGF0ZT5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyBjcmVhdGVTU1JBcHAsIGgsIG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IHJlbmRlclRvU3RyaW5nIH0gZnJvbSAndnVlL3NlcnZlci1yZW5kZXJlcidcbmltcG9ydCB7IEVsVGFicywgRWxUYWJQYW5lIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCBhcHAgPSBjcmVhdGVTU1JBcHAoe1xuICByZW5kZXIoKSB7XG4gICAgY29uc3QgdGFiUGFuZXMgPSBbJ2ZvbycsICdiYXInXS5tYXAoKHYpID0+XG4gICAgICBoKFxuICAgICAgICBFbFRhYlBhbmUsXG4gICAgICAgIHtcbiAgICAgICAgICBsYWJlbDogdixcbiAgICAgICAgICBuYW1lOiB2LFxuICAgICAgICB9LFxuICAgICAgICB7XG4gICAgICAgICAgZGVmYXVsdDogKCkgPT4gdixcbiAgICAgICAgfVxuICAgICAgKVxuICAgIClcblxuICAgIHJldHVybiBoKFxuICAgICAgRWxUYWJzLFxuICAgICAge1xuICAgICAgICBtb2RlbFZhbHVlOiAnYmFyJyxcbiAgICAgIH0sXG4gICAgICB7IGRlZmF1bHQ6ICgpID0+IHRhYlBhbmVzIH1cbiAgICApXG4gIH0sXG59KVxuXG5vbk1vdW50ZWQoKCkgPT4ge1xuICByZW5kZXJUb1N0cmluZyhhcHApLnRoZW4oKGh0bWwpID0+IHtcbiAgICBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCcjY29udGFpbmVyJykuaW5uZXJIVE1MID0gaHRtbFxuICB9KVxufSlcbjwvc2NyaXB0PlxuIiwic3JjL1BsYXlncm91bmRNYWluLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQXBwIGZyb20gJy4vQXBwLnZ1ZSdcbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEFwcCAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0xNTE4My1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCIsXG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL3Z1ZS9kaXN0L3Z1ZS5ydW50aW1lLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJ2dWUvc2VydmVyLXJlbmRlcmVyXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3NlcnZlci1yZW5kZXJlckBsYXRlc3QvZGlzdC9zZXJ2ZXItcmVuZGVyZXIuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcIkB2dWUvcnVudGltZS1kb21cIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcIkB2dWUvcnVudGltZS1jb3JlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtY29yZUBsYXRlc3QvZGlzdC9ydW50aW1lLWNvcmUuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcIkB2dWUvcmVhY3Rpdml0eVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9yZWFjdGl2aXR5QGxhdGVzdC9kaXN0L3JlYWN0aXZpdHkuZXNtLWJ1bmRsZXIuanNcIlxuICB9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsInNyYy9lbGVtZW50LXBsdXMuanMiOiJpbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5pbXBvcnQgRWxlbWVudFBsdXMgZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5sZXQgaW5zdGFsbGVkID0gZmFsc2VcbmF3YWl0IGxvYWRTdHlsZSgpXG5cbmV4cG9ydCBmdW5jdGlvbiBzZXR1cEVsZW1lbnRQbHVzKCkge1xuICBpZiAoaW5zdGFsbGVkKSByZXR1cm5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoRWxlbWVudFBsdXMpXG4gIGluc3RhbGxlZCA9IHRydWVcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIGxvYWRTdHlsZSgpIHtcbiAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICBsaW5rLmhyZWYgPSAnaHR0cHM6Ly9wcmV2aWV3LTE1MTgzLWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzJ1xuICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdlcnJvcicsIHJlamVjdClcbiAgICBkb2N1bWVudC5ib2R5LmFwcGVuZChsaW5rKVxuICB9KVxufSIsIl9vIjp7InNob3dIaWRkZW4iOnRydWUsInN0eWxlU291cmNlIjoiaHR0cHM6Ly9wcmV2aWV3LTE1MTgzLWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzIn19)

### CODE REVIEW

**Review by tolking — APPROVED** [review]

**Review by tolking — APPROVED** [review]
LGTM.

**Comment by btea** [comment]
There are code conflicts that need to be resolved.

**Comment by makedopamine** [comment]
> There are code conflicts that need to be resolved.

Resolved

File: packages/components/tabs/src/tabs.tsx [hunk 1]
```diff
@@ -212,15 +219,19 @@ const Tabs = defineComponent({
...
-          <TabNav
-            ref={nav$}
-            currentName={currentName.value}
-            editable={props.editable}
-            type={props.type}
-            panes={panes.value}
-            stretch={props.stretch}
-            onTabClick={handleTabClick}
-            onTabRemove={handleTabRemove}
+          <TabNavRenderer
```
Comment (btea) [inline packages/components/tabs/src/tabs.tsx:222]: What is the purpose of this modification?

Comment (makedopamine) [reply packages/components/tabs/src/tabs.tsx:222]: It's to execute the creation of  `TabNav` VNode after `panels` were patched. Without it, the `panes.value` could be stale when creating `TabNav` VNode.

Comment (btea) [reply packages/components/tabs/src/tabs.tsx:222]: Thanks for the explanation.

**Review by btea — APPROVED** [review]

File: packages/components/tabs/src/tabs.tsx [hunk 2]
```diff
@@ -241,9 +252,8 @@ const Tabs = defineComponent({
             },
           ]}
         >
-          {...props.tabPosition !== 'bottom'
-            ? [header, panels]
-            : [panels, header]}
+          {panels}
+          {header}
```
[the remark this answers is not in the document]
Comment (makedopamine) [reply packages/components/tabs/src/tabs.tsx:256]: The reason had been mentioned in the above [comment](https://github.com/element-plus/element-plus/pull/15183#discussion_r1694922213).

**Comment by zhixiaotong** [comment]
@btea @tolking 这个pr导致了tabs的内容发生了改变，属于**breaking change**了。正常content在header下面，现在这个content在header上面了，然后又加了一个`flex-direction: column-reverse;`保证正常，但影响了原有的项目，升级新版本后，发现项目原有好多样式需要重写，影响蛮大的。

**Comment by tolking** [comment]
这个PR是属于breaking chang，所以才会在v2.8.0 合并。

对于普通用户来说影响应该挺小的。

**Comment by zhixiaotong** [comment]
也许我们应该在changelog中体现这些变更的细节。
我习惯性在tabs下使用纵向flex布局，在升级之前一般会先看changlog，但问题发生才回溯代码发现是某些原因，之前也有几次升级出现过类似的问题。
我以前看过一些issue风评不好，这会导致用户在选择升级版本变得更加不确定。

**Comment by btea** [comment]
@zhixiaotong 多谢反馈，后续我们会尝试优化这方面的工作。

**Pull request merged**